src/execution/operator/aggregate/physical_streaming_window.cpp

```diff
@@ -118,6 +101,98 @@ class StreamingWindowState : public OperatorState {
 		Vector addresses;
 	};
 
+	struct LeadLagState {
+		static constexpr idx_t MAX_BUFFER = STANDARD_VECTOR_SIZE;
+
+		static bool ComputeOffset(ClientContext &context, BoundWindowExpression &wexpr, int64_t &offset) {
+			offset = 1;
+			if (wexpr.offset_expr) {
+				if (wexpr.offset_expr->HasParameter() || !wexpr.offset_expr->IsFoldable()) {
+					return false;
+				}
+				auto offset_value = ExpressionExecutor::EvaluateScalar(context, *wexpr.offset_expr);
+				if (offset_value.IsNull()) {
+					return false;
+				}
+				Value bigint_value;
+				if (!offset_value.DefaultTryCastAs(LogicalType::BIGINT, bigint_value, nullptr, false)) {
+					return false;
+				}
+				offset = bigint_value.GetValue<int64_t>();
+			}
+
+			//	We can only support negative LEAD values and positive LAG values
+			if (wexpr.type == ExpressionType::WINDOW_LEAD) {
+				offset = -offset;
+			}
+			return 0 <= offset && idx_t(offset) < MAX_BUFFER;
+		}
+
+		static bool ComputeDefault(ClientContext &context, BoundWindowExpression &wexpr, Value &result) {
+			if (!wexpr.default_expr) {
+				result = Value(wexpr.return_type);
+				return true;
+			}
+
+			if (wexpr.default_expr && (wexpr.default_expr->HasParameter() || !wexpr.default_expr->IsFoldable())) {
+				return false;
+			}
+			auto dflt_value = ExpressionExecutor::EvaluateScalar(context, *wexpr.default_expr);
+			return dflt_value.DefaultTryCastAs(wexpr.return_type, result, nullptr, false);
+		}
+
+		LeadLagState(ClientContext &context, BoundWindowExpression &wexpr)
+		    : wexpr(wexpr), curr(wexpr.return_type), prev(wexpr.return_type, MAX_BUFFER),
+		      temp(wexpr.return_type, MAX_BUFFER) {
+			ComputeOffset(context, wexpr, offset);
+			ComputeDefault(context, wexpr, dflt);
+			prev.Reference(dflt);
+			prev.Flatten(MAX_BUFFER);
+		}
+
+		void Execute(ExecutionContext &context, DataChunk &input, Vector &result) {
+			ExpressionExecutor executor(context.client, *wexpr.children[0]);
+			executor.ExecuteExpression(input, curr);
+
+			//	Copy prev[MAX_BUFFER-offset, MAX_BUFFER] => result[0, offset]
+			idx_t source_count = MAX_BUFFER;
```

Comment:
Since the lead/lag are likely to be small (in my experience at least, lead/lag of 1 is the most common) - perhaps we can buffer only the required amount of rows for the lead/lag instead of keeping around and shifting an entire vector at all times?